src/traced/probes/user_list/user_list_data_source.cc

```diff
@@ -0,0 +1,148 @@
+
+#include "perfetto/ext/tracing/core/trace_writer.h"
+#include "protos/perfetto/trace/trace_packet.pbzero.h"
+
+using perfetto::protos::pbzero::AndroidUserListConfig;
+
+namespace perfetto {
+
+// static
+const ProbesDataSource::Descriptor UserListDataSource::descriptor = {
+    /*type*/ "android.user_list",
+    /*flags*/ Descriptor::kFlagsNone,
+    /*fill_descriptor_func*/ nullptr,
+};
+
+UserListDataSource::UserListDataSource(const DataSourceConfig& ds_config,
+                                       TracingSessionID session_id,
+                                       std::unique_ptr<TraceWriter> writer)
+    : ProbesDataSource(session_id, &descriptor), writer_(std::move(writer)) {
+  AndroidUserListConfig::Decoder cfg(ds_config.user_list_config_raw());
+  for (auto type = cfg.user_type_filter(); type; ++type) {
+    user_type_filter_.emplace((*type).ToStdString());
+  }
+}
+
+void UserListDataSource::Start() {
+  auto trace_packet = writer_->NewTracePacket();
+  auto* user_list_packet = trace_packet->set_user_list();
+  int error_code = 0;
+
+  base::ScopedFstream fs(fopen("/data/system/users/user.list", "r"));
+  if (!fs) {
+    error_code = errno;  // Capture errno from fopen failure
+    PERFETTO_ELOG("Failed to open user.list: %s", strerror(error_code));
+    user_list_packet->set_error(error_code);
+    trace_packet->Finalize();
+    writer_->Flush();
+    return;
+  }
+
+  error_code = ParseUserListStream(user_list_packet, fs, user_type_filter_);
+  if (error_code != 0) {
+    PERFETTO_DLOG("Failed to parse user.list content: %s",
+                  strerror(error_code));
+    user_list_packet->set_error(error_code);
+  }
+
+  if (ferror(*fs)) {
+    error_code = errno;  // Capture errno from stream error
+    PERFETTO_ELOG("Error reading user.list: %s", strerror(error_code));
+    // Overwrite any previous error, as read error is more fundamental.
+    user_list_packet->set_error(error_code);
+  }
+
+  trace_packet->Finalize();
+  writer_->Flush();
+}
+
+void UserListDataSource::Flush(FlushRequestID, std::function<void()> callback) {
+  // Flush is no-op. We flush after the first write.
+  callback();
+}
+
+UserListDataSource::~UserListDataSource() = default;
+
+// parser
+
+// Returns 0 on success, EPROTO on parsing failure.
+int ReadUserListLine(char* line, User* user) {
+  line[strcspn(line, "\r\n")] = 0;
+  size_t idx = 0;
+  for (base::StringSplitter str_splitter(line, ' '); str_splitter.Next();) {
+    switch (idx) {
+      case 0:
+        user->type = std::string(str_splitter.cur_token(),
+                                 str_splitter.cur_token_size());
+        break;
+      case 1: {
+        std::optional<int32_t> cur_uid =
+            base::CStringToInt32(str_splitter.cur_token());
+        if (cur_uid.has_value()) {
+          user->uid = cur_uid.value();
+        } else {
+          PERFETTO_DLOG("Failed to parse user.list cur_uid.");
+          return EPROTO;  // Protocol error
```

Comment:
can you just return -1? you don't seem to make any use of the specific error. ANd also is a bit odd to see this.
EPROTO is a Linux-ism and will cause compilation errors on people using this in other OSes (i'm thinking of QNX used by car manufacturers)